src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs

```diff
@@ -343,12 +347,52 @@ private static bool IsSelfClosing(Block beginTagBlock)
                 }
             }
 
-            result.AttributeValueNode = ConvertToMarkupAttributeBlock(block, result.IsBoundNonStringAttribute);
+            var isFirstSpan = true;
+
+            result.AttributeValueNode = ConvertToMarkupAttributeBlock(
+                block,
+                spanBuilder =>
+                {
+                    // If the attribute was requested by a tag helper but the corresponding property was not a
+                    // string, then treat its value as code. A non-string value can be any C# value so we need
+                    // to ensure the SyntaxTreeNode reflects that.
+                    if (result.IsBoundNonStringAttribute)
+                    {
+                        // For bound non-string attributes, we'll only allow a transition span to appear at the very
+                        // beginning of the attribute expression. All later transitions would appear as code so that
+                        // they are part of the generated output. E.g.
+                        // key="@value" -> MyTagHelper.key = value
+                        // key=" @value" -> MyTagHelper.key =  @value
+                        // key="1 + @case" -> MyTagHelper.key = 1 + @case
+                        // key="@int + @case" -> MyTagHelper.key = int + @case
+                        // key="4 + @(a + b)" -> MyTagHelper.key = 4 + @a + b
```

Comment:
Does the comment correctly describe the current code? Looks as if the last case results in `4 + @(a + b)`. May also be worth mentioning `@(a + b)` -> `(a + b)`.